api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,268 @@ package messagebuffer
-	defer messageBuffer.Unlock()
-	messages := messageBuffer.buffer[clientIDStr]
-	messageBuffer.buffer[clientIDStr] = nil
-	return messages
+
+	msgBuf.Lock()
+	defer msgBuf.Unlock()
+
+	treeMsg, ok := msgBuf.buffer[clientIDStr]
+	if !ok {
+		return nil
+	}
+
+	relayMessage := make([]*pb.Relay, 0, treeMsg.Size())
+	for node := treeMsg.Left(); node != nil; node = treeMsg.Left() { // delete them from index tree
+
+		mNode := node.Value.(*msgNode)
+		relayMessage = append(relayMessage, mNode.msg)
+		treeMsg.Remove(mNode.createId)
+		msgBuf.totalBufSize -= mNode.GetSize()
+
+		msgBuf.treeCreate.Remove(mNode.createId)
+		msgBuf.treeExpire.Remove(mNode.expireId)
+
+	}
+
+	return relayMessage
+}
+
+// remove expired message
+func (msgBuf *MessageBuffer) removeExpired() (removedCount int) {
+	now := time.Now().UnixMilli() + 1 // plus 1 milli-second
+	nowId := now * maxSeqId
+
+	removedCount = 0
+
+	msgBuf.Lock()
+	defer msgBuf.Unlock()
+
+	for minNode := msgBuf.treeExpire.Left(); minNode != nil; minNode = msgBuf.treeExpire.Left() {
+		idxNode := minNode.Value.(*indexNode)
+
+		if idxNode.expireId < nowId { // expired
+			v, ok := msgBuf.buffer[idxNode.clientID].Get(idxNode.createId)
+			if ok { // remove from buffer
+				mNode := v.(*msgNode)
+				msgBuf.buffer[idxNode.clientID].Remove(idxNode.createId)
+				msgBuf.totalBufSize -= mNode.GetSize()
+			}
+			// remove from index tree
+			msgBuf.treeExpire.Remove(idxNode.expireId)
+			msgBuf.treeCreate.Remove(idxNode.createId)
+			removedCount++
+		} else {
+			break
+		}
+	}
+
+	return
+}
+
+// remove oldest message
+func (msgBuf *MessageBuffer) removeOldest() {
+
+	minNode := msgBuf.treeCreate.Left()
+	if minNode == nil { // the tree is empty
+		return
+	}
+
+	msgBuf.Lock()
+	defer msgBuf.Unlock()
+
+	createId := minNode.Key.(int64)
+	idxNode := minNode.Value.(*indexNode)
+	treeMsg, ok := msgBuf.buffer[idxNode.clientID]
+	if ok {
+		v, ok := treeMsg.Get(createId)
+		if ok {
+			treeMsg.Remove(createId)
+			mNode := v.(*msgNode)
+			msgBuf.totalBufSize -= mNode.GetSize()
+		}
+
+		msgBuf.treeExpire.Remove(idxNode.expireId)
+		msgBuf.treeCreate.Remove(createId)
+	}
+
+}
+
+// clear all cached msg and index
+func (msgBuf *MessageBuffer) ClearAll() {
+	msgBuf.Lock()
+	defer msgBuf.Unlock()
+
+	for _, treeMsg := range msgBuf.buffer {
+		treeMsg.Clear()
+	}
+	msgBuf.treeCreate.Clear()
+	msgBuf.treeExpire.Clear()
+}
+
+// monitor message buffer usage, and remove expired message .
+func (msgBuf *MessageBuffer) BufferMonitor() {
+	if config.Parameters.ClientMsgCacheSize <= 0 {
+		log.Error("config ClientMsgCacheSize is not set properly. ")
+	}
+
+	for {
+		time.Sleep(MonitorInterval * time.Second)
+
+		// delete expired msg
+		msgBuf.removeExpired()
+
+		msgBuf.Lock()
+		totalBufSize := msgBuf.totalBufSize
+		msgBuf.Unlock()
+
+		// if over size, repeat until totalBufSize less than config.ClientMsgCacheSize
+		for config.Parameters.ClientMsgCacheSize > 0 && totalBufSize > int(config.Parameters.ClientMsgCacheSize) {
```

Comment:
`totalBufSize` might be changed between L242 and L246. We can remove lock inside `removeExpired()` and `removeOldest()`, lock before `removeExpired()` call, and unlock after the for loop.

Reply:
yes. removed lock from removeExpired() and removeOldest(). and add lock in the loop. I just worry later if some other people use removeOldest() or removeExpired() seperately and forget use lock and unlock. :)